src/bosh-director/lib/bosh/director/jobs/update_deployment.rb

```diff
@@ -89,6 +89,60 @@ def prepare_deployment
         end
       end
 
+      # unused_subnets returns all the subnets in the iaas that are not being used any more
+      def unused_subnets(new_subnets, old_subnets)
+        old_subnets.select do |subnet|
+          new_subnets.find { |e| e.name == subnet.name }.nil?
+        end
+      end
+
+      def used_cloud_config_state(deployment)
+        latest_cloud_configs = Models::Config.latest_set_for_teams('cloud', *deployment.teams).map(&:id).sort
+        if deployment.cloud_configs.empty?
+          'none'
+        elsif deployment.cloud_configs.map(&:id).sort == latest_cloud_configs
+          'latest'
+        else
+          'outdated'
+        end
+      end
+
+      # removes a subnet model in the iaas and from the database
+      def delete_subnet(cloud_factory, subnet)
+        cpi = cloud_factory.get(subnet.cpi)
+        begin
+          @logger.info("deleting unused subnet #{subnet.name}")
+          cpi.delete_network(subnet.cid)
+        rescue StandardError => e
+          # failing to delete the subnet in the iaas shouldn't fail the deployment
+          @logger.error("failed to delete subnet #{subnet.name}: #{e.message}")
```

Comment:
It seems unusual to attempt to remove the subnet, and then if it fails, remove it from the bosh database, and never retry deleting it.  

Maybe it shouldn't destroy it from the bosh database in the failure state, so that the next time it does a cleanup, bosh can try again.

Reply:
Does the CPI result give a hint of whether the IaaS isn't responding versus the subnet is already deleted?  Maybe that could be the difference of whether the subnet db model should be purged or not.

Reply:
Currently, the CPI doesn't give any hints regarding why a failure could have occurred. I was initially thinking of removing the subnet from the IaaS as a best effort attempt. Meaning that Bosh will attempt to remove the subnet, if it fails for any reason, Bosh would remove this subnet from the database and stop managing it and then it will be left to the cloud operator to clean things up. But you are right, another approach could be to leave it in the database. This way, Bosh will attempt to remove it again the next time the deployment is updated. I have no strong opinions either way.

Reply:
One problem I see in the second approach is:
1- deployment created with a managed network with two subnets
2- cloud config updated, and the second subnet is removed from the config
3- IaaS returns an error "can not delete subnet" and the deployment update proceeds
4- The cloud operator manually cleans up the no longer used subnet
5- any further update of the deployment will attempt to remove the subnet from the iaas (which no longer exists) but will fail
In this scenario, the subnet will never be removed from the database.

I have another PR ready to be submitted for this approach if this is the desired behavior.

Reply:
Hi @kreamyx,

Sorry we haven't responded in a while.

With regards to the second approach, is it possible for us to detect whether or not the subnet is already deleted? If so, then when deleting a subnet, if the subnet does not exist we can safely move on. This would be similar to the behavior for VMs.

Thoughts?

Also it appears that this PR has gotten somewhat stale due to our inactivity. Is there any chance you could rebase this on PR with the master branch?

Best,

@jfmyers9 && @belinda-liu, CF BOSH

Reply:
Hi @jfmyers9 and @belinda-liu 

Thanks for reviewing the PR.

Currently, there is no way to find out if a subnet doesn't exist in the IAAS (we currently only support `create_network` and `delete_network` in the CPI). But I think the idea of having an additional `has_network` in the future would be very useful in this case (and perhaps other cases as well), and it will make it consistent with the VM-related CPI methods as well.

I have also rebased my branch on top of master.

Thanks,
Karim

Reply:
Since we do not currently have a `has_network` method, would it be possible to distinguish between when the deletion fails due to the fact that the subnet no longer exists? This would allow BOSH to safely retry even if the operator has already manually cleaned up the IaaS. 
If that's possible, then let's keep the subnet in the database so that BOSh can retry when deleting fails.